File: LeetCode/1-100/11. Container With Most Water/Solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <unordered_map>

using namespace std;

class Solution {
public:
    int maxArea(vector<int>& height) {
        unordered_map<int, int> tracker;
        int n = height.size();
        unordered_map<int, int>::iterator pointer;
        int maxWater = 0;
        for (int i = 0; i < n; ++i) {
            pointer = tracker.find(height[i]);
            if (pointer != tracker.end()) {
                maxWater = max(maxWater, (i - pointer->second) * height[i]);
            }
            pointer->second = i;
        }
    }
};

int main() {
    std::cout << "Hello, World!" << std::endl;
    return 0;
}

File: LeetCode/1-100/11. Container With Most Water/Solution.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        low, high, max_water, prev = 0, len(height) - 1, 0, -1
        while low < high:
            max_water = max(max_water, min(height[low], height[high]) * (high - low))
            if height[low] < height[high]:
                prev = height[low]
                while low < high and height[low] <= prev: low += 1
            else:
                prev = height[high]
                while low < high and height[high] <= prev: high -= 1
        return max_water